File: client/java/src/main/java/com/be/mesh/client/mpc/GenericInspector.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.mpc;

import com.be.mesh.client.cause.NotFoundException;
import lombok.AllArgsConstructor;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.util.Map;

/**
 * @author coyzeng@gmail.com
 */
@AllArgsConstructor
public class GenericInspector implements Inspector {

    private final String name;

    @Override
    public String getSignature() {
        return this.name;
    }

    @Override
    public Class<?> getType() {
        return Object.class;
    }

    @Override
    public String getName() {
        return this.name;
    }

    @Override
    public <T extends Annotation> T getAnnotation(Class<T> kind) {
        return null;
    }

    @Override
    public Class<?> getReturnType() {
        return Map.class;
    }

    @Override
    public Type getReturnGenericType() {
        return Map.class;
    }

    @Override
    public Class<?>[] getExceptionTypes() {
        return new Class[0];
    }

    @Override
    public Object invoke(Object obj, Object... args) throws ReflectiveOperationException {
        throw new NotFoundException("Generic execution cant serve %s as mps.", Mesh.context().getUrn());
    }
}

File: client/golang/mpc/generic_inspector.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package mpc

import (
	"context"
	"github.com/opendatav/mesh/client/golang/macro"
)

func init() {
	var _ macro.Inspector = new(GenericInspector)
}

type GenericInspector struct {
	Name string
	Args map[string]interface{}
}

func (that *GenericInspector) Call(ctx context.Context, proxy interface{}, method macro.Inspector, args ...interface{}) (interface{}, error) {
	return nil, nil
}

func (that *GenericInspector) String() string {
	return that.Name
}

func (that *GenericInspector) Signature() string {
	return that.Name
}

func (that *GenericInspector) GetDeclaredKind() interface{} {
	return (*interface{})(nil)
}

func (that *GenericInspector) GetName() string {
	return that.Name
}

func (that *GenericInspector) GetTName() string {
	return that.Name
}

func (that *GenericInspector) GetIntype() macro.Parameters {
	parameters := &GenericParameters{}
	return parameters
}

func (that *GenericInspector) GetRetype() macro.Returns {
	var returns GenericReturns
	return &returns
}

func (that *GenericInspector) NewInbound() macro.Parameters {
	return GenericParameters{}
}

func (that *GenericInspector) NewOutbound() macro.Returns {
	return GenericReturns{}
}

func (that *GenericInspector) GetMPS() macro.MPS {
	return &macro.MPSAnnotation{Meta: &macro.Stt{Name: that.Name}}
}

func (that *GenericInspector) GetMPI() macro.MPI {
	return &macro.MPIAnnotation{Meta: &macro.Rtt{Name: that.Name}}
}

func (that *GenericInspector) GetSPI() macro.SPI {
	return &macro.SPIAnnotation{Attribute: &macro.Att{Name: that.Name}}
}

func (that *GenericInspector) GetBinding() []macro.Binding {
	return []macro.Binding{&macro.BindingAnnotation{Binding: &macro.Btt{Topic: that.Name}}}
}
